proseco/jupiter.py

```diff
@@ -0,0 +1,320 @@
+from cheta.comps import ephem_stk
+from cxotime import CxoTime, CxoTimeLike
+from Quaternion import QuatLike
+
+from proseco import characteristics_jupiter
+from proseco.characteristics_jupiter import JupiterPositionTable
+
+if TYPE_CHECKING:
+    from proseco.core import StarsTable
+
+
+def date_is_excluded(date: CxoTimeLike) -> bool:
+    """
+    Check if the given date is in the list of excluded dates for Jupiter observations.
+
+    If the date is within any of the excluded date ranges, Jupiter checks do not apply,
+    and this function returns True.
+
+    Parameters
+    ----------
+    date : CxoTimeLike
+        The date to check.
+
+    Returns
+    -------
+    bool
+        True if the date is in an excluded range, False otherwise.
+    """
+    date_str = CxoTime(date).date
+    # Exclude dates are in CXO date format and we compare as strings. This prevents a
+    # warning from pyerfa about dubius dates in the future where leap seconds may be
+    # uncertain.
+    for exclude_date in characteristics_jupiter.exclude_dates:
+        if exclude_date["start"] <= date_str <= exclude_date["stop"]:
+            return True
+    return False
+
+
+def get_jupiter_position(
+    date: CxoTimeLike,
+    duration: float,
+    att: QuatLike,
+) -> JupiterPositionTable:
+    """
+    Get the position of Jupiter on the ACA CCD.
+
+    Parameters
+    ----------
+    date : CxoTimeLike
+        The start date of the observation (acquisition time)
+    duration : float
+        The duration of the observation in seconds.
+    att : Quaternion or Quat-compatible
+        The attitude Quaternion.
+
+    Returns
+    -------
+    JupiterPositionTable
+        A table with columns 'time', 'row', 'col' for the times when Jupiter
+        is on the CCD. If Jupiter is never on the CCD this is a table with zero rows.
+    """
+    date0 = CxoTime(date)
+    # dates is a 1-d array with a minimum of 2 points (the end points)
+    dates = CxoTime.linspace(date0, date0 + duration * u.s, step_max=1000 * u.s)
+    times = dates.secs
+
+    chandra_ephem = ephem_stk.get_ephemeris_stk(start=dates[0], stop=dates[-1])
+
+    # Get Jupiter positions using chandra_aca.planets
+    ephem = {
+        key: np.interp(times, chandra_ephem["time"], chandra_ephem[key])
+        for key in ["x", "y", "z"]
+    }
+    # shape (len(dates), 3), units km
+    pos_earth = planets.get_planet_barycentric("earth", dates)
+
+    chandra_eci = np.array(
+        [
+            ephem["x"] / 1000,  # convert m from get_ephemeris_stk to km,
+            ephem["y"] / 1000,
+            ephem["z"] / 1000,
+        ]
+    ).transpose()
+    eci = planets.get_planet_eci("jupiter", dates, pos_observer=pos_earth + chandra_eci)
+
+    # Convert ECI position to RA, Dec => yag, zag => row, col
+    ra, dec = eci_to_radec(eci)
+    yag, zag = radec_to_yagzag(ra, dec, att)
+    row, col = yagzag_to_pixels(yag, zag, allow_bad=True)
+
+    lim0 = -512
+    lim1 = 511
```

Comment:
I don't remember why this isn't 512 (typo?), but @taldcroft  should it be a few pixels bigger than the CCD for this check?  We're in the weeds but I figure I need to fix this anyway.

Reply:
Thanks!